### PR TITLE
Documentation: add development guide

### PR DESCRIPTION
**Pull request by johnwcallahan**
## Description

This adds a development guide with three entries:

1. Repository Structure 
2. Component Structure 
3. Managing State

There are more guides to be written but this is a decent start.

Still to come (in a separate PR):

4. Fetching Data 
5. Component Performance
6. Writing Unit Tests
7. Mocking Data 
8. Using the Local Dev Tools
9. Writing a Feature Flag
10. Installing and Updating Dependencies

### CODE REVIEW

File: docs/development-guide/01-repository-structure.md [hunk 1]
```diff
@@ -0,0 +1,162 @@
...
+
+Any files relevant to the entire project or repo should be included at the root level. Files belonging to a specific package belong in `/packages/<package_name>`.
+
+## api-v4
+
+The application code for the api-v4 package lives in `/src`. From there, each section of the API has its own directory, e.g. `/account`, `/domains`, `/linodes`, etc.
+
+Each subdirectory typically has:
+
+- A file with the API request methods (e.g. `account.ts`)
```
Comment (tiffwong) [inline docs/development-guide/01-repository-structure.md:21]: Should be a `.` at the end of the sentence so it's consistent with the other bullets

File: docs/development-guide/01-repository-structure.md [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+- `configs.ts`
+- `disks.ts`
+- ...etc.
+
+There are also a few key files in the /src directory of the api-v4 package:
+
+- constants.ts (package-wide constants, like API_ROOT)
+- index.ts (exports everything from the package)
+- request.ts (the nuts and bolts of making an API request, utilizing axios for XHR requests)
+- types.ts (package-wide TypeScript types and interfaces, like APIError.
```
Comment (tiffwong) [inline docs/development-guide/01-repository-structure.md:38]: Should replace the period at the end with a parenthesis

File: docs/development-guide/01-repository-structure.md [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+- **App.tsx**
+  - the container for the React application
+- **constants.ts**
+  - app-wide constants, either read from the environment or defined directly
+- **events.ts and eventsPolling.ts**
+  - events polling logic (events come from the /account/events endpoint and drive many behaviors throughout the app)
+- **featureFlags.ts**
+  - TypeScript types and interfaces corresponding to the feature flags we expect from LaunchDarkly (our third-party feature flagging service)
+- **index.tsx**
+  - the entry point for the React app. Uses `ReactDOM.render()` to render the app to ./public/index.html.
```
Comment (tiffwong) [inline docs/development-guide/01-repository-structure.md:82]: Remove the period after `./public/index.html`

File: docs/development-guide/01-repository-structure.md [hunk 4]
```diff
@@ -0,0 +1,162 @@
...
+- **/mocks**
+  - code powering the mock service worker, which allows easy mocking of the API during development
+- **/queries**
+  - code that uses react-query and the api-v4 package to fetch data from the API
+- **/store**
+  - redux actions, reducers, thunks
+- **/types**
+  - miscellaneous type declarations
+- **/utilities**
+  - miscellaneous reusable utility functions, e.g. “capitalize.ts”.
```
Comment (tiffwong) [inline docs/development-guide/01-repository-structure.md:123]: Remove the period here as well

File: docs/development-guide/01-repository-structure.md [hunk 5]
```diff
@@ -0,0 +1,162 @@
...
+
+## Where to go for common tasks
+
+### "I want to…"
+
+**“... consume data from a new API endpoint”**
+
+1. Add the request handler and types in `/packages/api-v4`
+2. If appropriate, add the schema to `/packages/validation`
+3. Add a React Query hook in `/packages/manager/src/queries`.
```
Comment (tiffwong) [inline docs/development-guide/01-repository-structure.md:133]: Remove the period here

File: docs/development-guide/01-repository-structure.md [hunk 6]
```diff
@@ -0,0 +1,162 @@
...
+4. Use the new hook in a function component
+
+**“... create a new, reusable component”**
+
+1. Add a new directory (upper camel case) in `/packages/manager/src/components`.
```
Comment (tiffwong) [inline docs/development-guide/01-repository-structure.md:138]: Maybe remove this period as well

File: docs/development-guide/01-repository-structure.md [hunk 7]
```diff
@@ -0,0 +1,162 @@
...
+2. Author the React component, unit tests, and Storybook stories across four files:
+   a. `<ComponentName>.tsx` (React component code)
+   b. `<ComponentName>.test.tsx` (unit tests for the React component)
+   c. `<ComponentName>.stories.tsx` (Storybook stories for the React component
+   d. `index.ts` (export the React component)
+
+**“...implement a new feature in an existing section of the app”**
```
Comment (tiffwong) [inline docs/development-guide/01-repository-structure.md:145]: Probably want to make the spacing after the ellipses consistent as well

File: docs/development-guide/01-repository-structure.md [hunk 8]
```diff
@@ -0,0 +1,162 @@
...
+
+1. Find the appropriate code in `/packages/manager/src/features`
+2. Modify existing components, add new components, etc.
+
+**“...create an entirely new section of the app”**
+
+1. Add a <Route /> in `src/MainContenxt.tsx`
```
Comment (tiffwong) [inline docs/development-guide/01-repository-structure.md:152]: Wrap `<Route />` in ticks

File: docs/development-guide/01-repository-structure.md [hunk 9]
```diff
@@ -0,0 +1,162 @@
...
+2. Add new section to the Primary Nav in `src/components/PrimaryNav`
+3. Add a directory (upper camel case) in /packages/manager/src/features`
+4. Add new feature components and corresponding tests
+
+**“...change the styling of a specific feature component”**
+
+1. Find where the styles are defined for the component you want to modify
+   a. They are likely defined in the feature component’s file, e.g. `src/features/<MyFeature>/<SomeComponent>.tsx`.
+2. Avoid making changes in `src/index.css`, `src/themeactory.ts`, and `src/themes.ts` unless you are intentionally making a global styling change.
```
Comment (tiffwong) [inline docs/development-guide/01-repository-structure.md:161]: Should be `themeFactory`

File: docs/development-guide/03-component-state.md [hunk 10]
```diff
@@ -0,0 +1,55 @@
...
+### Redux
+
+Historically Redux was use heavily for state management and data storage, including almost all data fetched from the API.
+
+Currently, some API data is still stored in Redux, but data fetching in general is being migrated to React Query. Redux is **still used for some global state**.
+
+- Redux code lives in `src/store`.
+- Each slice of state has its own directory in `src/store` which includes actions and the reducer.
+- Reducers are combined via `combineReducers` in `src/store/index`.
+- Components consume Redux state by using the standard react-redux `connect` HOC, though it's common to write a custom, reusable container HOC in `src/containers`.for each state slice in
```
Comment (tiffwong) [inline docs/development-guide/03-component-state.md:49]: I think there's an accidental period in "`src/containers`.for"

**Review by tiffwong — APPROVED** [review]
Left some cleanup but this is great!

**Review by DevDW — APPROVED** [review]
Great information to have documented 🚀

**Review by HanaXu — APPROVED** [review]
🧠 🚀

**Comment by johnwcallahan** [comment]
Thanks for the cleanup, @tiffwong! Addressed your comments.

**Pull request merged**